File: advertising_sdk/src/main/java/com/adadapted/android/sdk/core/addit/PayloadClient.java
package com.adadapted.android.sdk.core.addit;

import com.adadapted.android.sdk.config.EventStrings;
import com.adadapted.android.sdk.core.atl.AddToListItem;
import com.adadapted.android.sdk.core.concurrency.ThreadPoolInteractorExecuter;
import com.adadapted.android.sdk.core.device.DeviceInfo;
import com.adadapted.android.sdk.core.device.DeviceInfoClient;
import com.adadapted.android.sdk.core.event.AppEventClient;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class PayloadClient {
    @SuppressWarnings("unused")
    private static final String LOGTAG = PayloadClient.class.getName();

    public interface Callback {
        void onPayloadAvailable(List<AdditContent> content);
    }

    private static PayloadClient instance;

    private static boolean deeplinkInProgress = false;
    private static final Lock lock = new ReentrantLock();

    public static synchronized void createInstance(final PayloadAdapter adapter) {
        if(instance == null) {
            instance = new PayloadClient(adapter);
        }
    }

    private static synchronized PayloadClient getInstance() {
        return instance;
    }

    public static synchronized void pickupPayloads(final Callback callback) {
        if(instance == null || deeplinkInProgress) {
            return;
        }

        lock.lock();
        try {
            if(deeplinkInProgress) {
                return;
            }
        }
        finally {
            lock.unlock();
        }

        ThreadPoolInteractorExecuter.getInstance().executeInBackground(new Runnable() {
            @Override
            public void run() {
                getInstance().performPickupPayload(callback);
            }
        });
    }

    public static synchronized void deeplinkInProgress() {
        lock.lock();
        try {
            deeplinkInProgress = true;
        }
        finally {
            lock.unlock();
        }
    }

    public static synchronized void deeplinkCompleted() {
        lock.lock();
        try {
            deeplinkInProgress = false;
        }
        finally {
            lock.unlock();
        }
    }

    static synchronized void markContentAcknowledged(final AdditContent content) {
        ThreadPoolInteractorExecuter.getInstance().executeInBackground(new Runnable() {
            @Override
            public void run() {
            final Map<String, String> eventParams = new HashMap<>();
            eventParams.put("payload_id", content.getPayloadId());
            eventParams.put("source", content.getAdditSource());

            AppEventClient.Companion.getInstance().trackSdkEvent(EventStrings.ADDIT_ADDED_TO_LIST, eventParams);

            if(content.isPayloadSource()) {
                getInstance().trackPayload(content, "delivered");
            }
            }
        });
    }

    public synchronized static void markContentItemAcknowledged(final AdditContent content,
                                                                final AddToListItem item) {
        ThreadPoolInteractorExecuter.getInstance().executeInBackground(new Runnable() {
            @Override
            public void run() {
            final Map<String, String> eventParams = new HashMap<>();
            eventParams.put("payload_id", content.getPayloadId());
            eventParams.put("tracking_id", item.getTrackingId());
            eventParams.put("item_name", item.getTitle());
            eventParams.put("source", content.getAdditSource());

            AppEventClient.Companion.getInstance().trackSdkEvent(EventStrings.ADDIT_ITEM_ADDED_TO_LIST, eventParams);
            }
        });
    }

    static synchronized void markContentDuplicate(final AdditContent content) {
        ThreadPoolInteractorExecuter.getInstance().executeInBackground(new Runnable() {
            @Override
            public void run() {
                final Map<String, String> eventParams = new HashMap<>();
                eventParams.put("payload_id", content.getPayloadId());

                AppEventClient.Companion.getInstance().trackSdkEvent(EventStrings.ADDIT_DUPLICATE_PAYLOAD, eventParams);

                if(content.isPayloadSource()) {
                    getInstance().trackPayload(content, "duplicate");
                }
            }
        });
    }

    static synchronized void markContentFailed(final AdditContent content,
                                               final String message) {
        ThreadPoolInteractorExecuter.getInstance().executeInBackground(new Runnable() {
            @Override
            public void run() {
                final Map<String, String> eventParams = new HashMap<>();
                eventParams.put("payload_id", content.getPayloadId());

                AppEventClient.Companion.getInstance().trackError(EventStrings.ADDIT_CONTENT_FAILED, message, eventParams);

                if(content.isPayloadSource()) {
                    getInstance().trackPayload(content, "rejected");
                }
            }
        });
    }

    public synchronized static void markContentItemFailed(final AdditContent content,
                                                          final AddToListItem item,
                                                          final String message) {
        ThreadPoolInteractorExecuter.getInstance().executeInBackground(new Runnable() {
            @Override
            public void run() {
            final Map<String, String> eventParams = new HashMap<>();
            eventParams.put("payload_id", content.getPayloadId());
            eventParams.put("tracking_id", item.getTrackingId());

            AppEventClient.Companion.getInstance().trackError(EventStrings.ADDIT_CONTENT_ITEM_FAILED, message, eventParams);
            }
        });
    }

    private final PayloadAdapter adapter;

    private PayloadClient(final PayloadAdapter adapter) {
        this.adapter = adapter;
    }

    private void performPickupPayload(final Callback callback) {
        DeviceInfoClient.Companion.getInstance().getDeviceInfo(new DeviceInfoClient.Callback() {
            @Override
            public void onDeviceInfoCollected(final DeviceInfo deviceInfo) {
                AppEventClient.Companion.getInstance().trackSdkEvent(EventStrings.PAYLOAD_PICKUP_ATTEMPT);

                adapter.pickup(deviceInfo, new PayloadAdapter.Callback() {
                    @Override
                    public void onSuccess(List<AdditContent> content) {
                        callback.onPayloadAvailable(content);
                    }
                });
            }
        });
    }

    private void trackPayload(final AdditContent content,
                              final String result) {
        final PayloadEvent event = new PayloadEvent(content.getPayloadId(), result);
        adapter.publishEvent(event);
    }
}

File: advertising_sdk/src/main/java/com/adadapted/android/sdk/core/addit/PayloadClient.kt
package com.adadapted.android.sdk.core.addit

import com.adadapted.android.sdk.config.EventStrings
import com.adadapted.android.sdk.core.atl.AddToListItem
import com.adadapted.android.sdk.core.concurrency.TransporterCoroutineScope
import com.adadapted.android.sdk.core.device.DeviceInfo
import com.adadapted.android.sdk.core.device.DeviceInfoClient
import com.adadapted.android.sdk.core.event.AppEventClient
import java.util.concurrent.locks.Lock
import java.util.concurrent.locks.ReentrantLock

class PayloadClient private constructor(private val adapter: PayloadAdapter, private val appEventClient: AppEventClient, private val transporter: TransporterCoroutineScope) {
    interface Callback {
        fun onPayloadAvailable(content: List<AdditContent>)
    }

    private fun performPickupPayload(callback: Callback) {
        DeviceInfoClient.getInstance().getDeviceInfo(object : DeviceInfoClient.Callback {
            override fun onDeviceInfoCollected(deviceInfo: DeviceInfo) {
                appEventClient.trackSdkEvent(EventStrings.PAYLOAD_PICKUP_ATTEMPT)

                adapter.pickup(deviceInfo, object: PayloadAdapter.Callback {
                    override fun onSuccess(content: List<AdditContent>) {
                        callback.onPayloadAvailable(content)
                    }
                })
            }
        })
    }

    private fun trackPayload(content: AdditContent, result: String) {
        val event = PayloadEvent(content.payloadId, result)
        adapter.publishEvent(event)
    }

    @Synchronized
    fun pickupPayloads(callback: Callback) {
        if (deeplinkInProgress) {
            return
        }
        transporter.dispatchToBackground {
            performPickupPayload(callback)
        }
    }

    @Synchronized
    fun deeplinkInProgress() {
        lock.lock()
        deeplinkInProgress = try {
            true
        } finally {
            lock.unlock()
        }
    }

    @Synchronized
    fun deeplinkCompleted() {
        lock.lock()
        deeplinkInProgress = try {
            false
        } finally {
            lock.unlock()
        }
    }

    @Synchronized
    fun markContentAcknowledged(content: AdditContent) {
        transporter.dispatchToBackground {
            val eventParams: MutableMap<String, String> = HashMap()
            eventParams[PAYLOAD_ID] = content.payloadId
            eventParams[SOURCE] = content.additSource
            appEventClient.trackSdkEvent(EventStrings.ADDIT_ADDED_TO_LIST, eventParams)
            if (content.isPayloadSource) {
                trackPayload(content, "delivered")
            }
        }
    }

    @Synchronized
    fun markContentItemAcknowledged(content: AdditContent, item: AddToListItem) {
        transporter.dispatchToBackground {
            val eventParams: MutableMap<String, String> = HashMap()
            eventParams[PAYLOAD_ID] = content.payloadId
            eventParams[TRACKING_ID] = item.trackingId
            eventParams[ITEM_NAME] = item.title
            eventParams[SOURCE] = content.additSource
            appEventClient.trackSdkEvent(EventStrings.ADDIT_ITEM_ADDED_TO_LIST, eventParams)
        }
    }

    @Synchronized
    fun markContentDuplicate(content: AdditContent) {
        transporter.dispatchToBackground {
            val eventParams: MutableMap<String, String> = HashMap()
            eventParams[PAYLOAD_ID] = content.payloadId
            appEventClient.trackSdkEvent(EventStrings.ADDIT_DUPLICATE_PAYLOAD, eventParams)
            if (content.isPayloadSource) {
                trackPayload(content, "duplicate")
            }
        }
    }

    @Synchronized
    fun markContentFailed(content: AdditContent, message: String) {
        transporter.dispatchToBackground {
            val eventParams: MutableMap<String, String> = HashMap()
            eventParams[PAYLOAD_ID] = content.payloadId
            appEventClient.trackError(EventStrings.ADDIT_CONTENT_FAILED, message, eventParams)
            if (content.isPayloadSource) {
                trackPayload(content, "rejected")
            }
        }
    }

    @Synchronized
    fun markContentItemFailed(content: AdditContent, item: AddToListItem, message: String) {
        transporter.dispatchToBackground {
            val eventParams: MutableMap<String, String> = HashMap()
            eventParams[PAYLOAD_ID] = content.payloadId
            eventParams[TRACKING_ID] = item.trackingId
            appEventClient.trackError(EventStrings.ADDIT_CONTENT_ITEM_FAILED, message, eventParams)
        }
    }

    companion object {
        private val LOGTAG = PayloadClient::class.java.name
        private const val PAYLOAD_ID = "payload_id"
        private const val TRACKING_ID = "tracking_id"
        private const val SOURCE = "source"
        private const val ITEM_NAME = "item_name"
        private lateinit var instance: PayloadClient
        private var deeplinkInProgress = false
        private val lock: Lock = ReentrantLock()

        fun getInstance(): PayloadClient {
            return instance
        }

        fun createInstance(adapter: PayloadAdapter, appEventClient: AppEventClient, transporter: TransporterCoroutineScope) {
            instance = PayloadClient(adapter, appEventClient, transporter)
        }
    }

}